### ISSUE TITLE
On all_gather in distributed training

### ISSUE THREAD

**Issue by seanzhuh** [mlfoundations/open_clip#1047]
Hi, I've encountered a problem possibly related with torch version.

In [this line of code](https://github.com/mlfoundations/open_clip/blob/main/src/open_clip/loss.py#L51), you use `torch.distributed.nn.all_gather` to gather all local features across all gpus. However, in torch version from 2.4.0 to the latest, the official doc has no `torch.distributed.nn.all_gather` but only `torch.distributed.all_gather` without the `nn` package anymore. And as I use `torch.distributed.all_gather` for distributed training, the loss will not go down and is higher than single-gpu training. This is because `torch.distributed.all_gather` does not keep tracking autograd, leadning to no gradient  calculated and flowed back to the model. Is there any workaround? Which torch version should I downgrade to?

**Comment by Zasder3** [comment]
Looking at the most recent version of PyTorch's code, it seems that [`nn.all_gather`](https://github.com/pytorch/pytorch/blob/17dbeb11db7afbab792ad76c24840c1552a0e76d/torch/distributed/nn/functional.py#L318) is still around. Though it's not in the docs it's still in the code.

**Comment by seanzhuh** [comment]
Thanks a lot. It turns out that using torch.distributed.nn.all_gather() will report torch.distributed has no 'nn' attribute but using from torch.distributed.nn import all_gather works fine.
